AI #1785: Add Invoice Detail and Billing Period datasets (#1900)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/invoice_detail/columns/referenceinvoiceid.md

# Reference Invoice ID

A Reference Invoice ID is an invoice-issuer-assigned identifier for an invoice that affects charges as stated on a previous invoice. This occurs typically in the context of credits, refunds, or corrections where an adjustment is applied to a specific previously-issued billing document. This ID allows for direct lineage between adjustments and the original billing records.

## Requirements

> ReferenceInvoiceId adheres to the following requirements:

Reviewer: ```suggestion
ReferenceInvoiceId MUST adhere to the following requirements:
```

* ReferenceInvoiceId MUST be of type String.
* ReferenceInvoiceId MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* ReferenceInvoiceId MUST NOT be null.

> * ReferenceInvoiceId MUST equal the [InvoiceId](#datasets.costandusage.invoiceid) of the original [*invoice*](#glossary:invoice) if it adjusts another invoice.

Reviewer 2: ```suggestion
* ReferenceInvoiceId MUST equal the [InvoiceId](#datasets.invoicedetail.invoiceid) of the original [*invoice*](#glossary:invoice) if it adjusts another invoice.
```

# Reference Invoice ID

A Reference Invoice ID is an invoice-issuer-assigned identifier for an invoice that affects charges as stated on a previous invoice. This occurs typically in the context of credits, refunds, or corrections where an adjustment is applied to a specific previously-issued billing document. This ID allows for direct lineage between adjustments and the original billing records.

## Requirements

ReferenceInvoiceId adheres to the following requirements:

* ReferenceInvoiceId MUST be present in an Invoice Detail [*FOCUS dataset*](#glossary:FOCUS-dataset).
* ReferenceInvoiceId MUST be of type String.
* ReferenceInvoiceId MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* ReferenceInvoiceId MUST NOT be null.

> * ReferenceInvoiceId MUST equal the [InvoiceId](#datasets.costandusage.invoiceid) of the original [*invoice*](#glossary:invoice) this invoice adjusts.

Reviewer 3: Can we word this in a clearer way? I think I understand the statement here but would like to know if it's possible to clarify. 

"this invoice adjusts." does this mean when its populated on a row in the dataset this row adjusts the previous invoice Id or is it this whole invoice adjusts the previous invoice. I think the current wording makes it sound like the later.

Author: How about:

```suggestion
* ReferenceInvoiceId MUST equal the [InvoiceId](#datasets.costandusage.invoiceid) of the original [*invoice*](#glossary:invoice) if it adjusts another invoice.
```

This then follows the pattern of the following normative requirement.

Author: @mike-finopsorg Bump.  Will commit and resolve if I do not hear from you by Feb 10 TF-1.

Author: No response received.  Committing and resolving.

> * ReferenceInvoiceId MUST equal the InvoiceId of the current invoice if it does not adjust another invoice.

Reviewer 3: Interesting choice, I assume this has been discussed and a choice made over making it NULL? if so feel free to close I'm not looking to force re-litigation, but if not possibly worth a discussion.

Author: I can bring it up with TF-1 tomorrow to get a pulse check.  In my past data modeling experience, it is ideal for this "reference" invoice ID to always be populated, so the consumer can more easily aggregate, rather than having to transform the column and fill it in for "original" invoices.

Author: Per @pydi-aws: There could be a situation where an invoice adjusts itself.  How would that surface here?  @ijurica will need to consider this scenario when working on the correction handling scenarios.

Author: Per @pydi-aws: there may be scenario of an adjustment invoice correcting multiple original invoices.  He will provide an example via AI #1902.

Author: Discussed in TF-1 over the last couple of weeks.  The group agreed to always populating it, and to limiting it to one and only one invoice.  We may down the road add a column that would facilitate a mapping to multiple invoices.  Happy to change this verbiage to specify that it adjusts a _single_ invoice.
